plutus-contract/src/Plutus/Contract/StateMachine/OnChain.hs

```diff
@@ -129,13 +129,13 @@ mkValidator (StateMachine step isFinal check threadToken) currentState input ptx
                 | otherwise ->
                     let txc =
                             newConstraints
-                                { txOwnOutputs=
+                                { txOwnOutputs =
                                     [ OutputConstraint
                                         { ocDatum = newData
                                           -- Check that the thread token value is still there
                                         , ocValue = newValue <> threadTokenValueInner threadToken (ownHash ptx)
                                         }
-                                    ]
+                                    | not (isZero newValue)]
```

Comment:
This was needed because of the GameStateMachine contract.

Here's an use case example which explains the mod:

- `Wallet1` locks 8 Ada to the script address
- `Wallet2` call the `guess` endpoint with the correct secret and wants the full 8 Ada.
- In `GameStateMachine.transition`, the `(Locked, Guess)` transition will generate a new `stateValue` of `mempty` (8 locked Ada - 8 requested Ada).
- Then, `StateMachine.runStep` for the `(Locked, Guess)` transition is executed and it creates an unbalanced tx containing an output of `mempty` Value to the script address. 
- `Wallet.Emulator.Wallet` will validate the balanced transaction and report the error `GameSMError (SMCContractError (WalletError (ValidationError (ScriptFailure (EvaluationError ["L1","Ld","S5","PT5"] "CekEvaluationFailure")))))`

The validation that is used is part of `StateMachine.OnChain.mkValidator` that I modified above. It will call `Ledger.Constraints.OnChain.checkOwnOutputConstraint`. Since the new `stateValue` is `mempty`, `getContinuingOutputs ctx` return an empty, list therefore `checkOwnOutputConstraint` returns `False` (because of `List.any`, thus the validation fails.

Therefore, my mod is simply to not create an output constraint is the new value is empty.

Reply:
But now it doesn't check for the thread token anymore. And what happens if the value is larger than zero but less than 2 Ada?

Reply:
> But now it doesn't check for the thread token anymore. 
You're right. I should replace it with `not (isZero (newValue <> threadTokenValueInner thread (ownHash ptx))`. I think this is valid. Note that the `runStep` and `mkStep` function will adjust the new state value to contain min of 2Ada is it's not empty. Also note, that I adjusted the `mkStep` function as well. Great job find that.

> And what happens if the value is larger than zero but less than 2 Ada?
So if `Wallet1` locks 8 Ada, and `Wallet2` correctly guesses the secret and want 7Ada, then the new state value will be 1Ada. `mkStep` will generate an adjusted contraint which adds the missing ada (output of 2Ada + thread token), and `runStep` will adjust the outputs of created unbalanced tx.
Makes sense?

Reply:
Alright, I managed to find a way to revert this change. What I needed to do was adapt the GameStateMachine contract such that when the locked funds are 0, then we change the state to final, so there will be no more txs.